%load_ext watermark
%watermark -a "Chibuzor Enyioko" -d -v -p numpy,pandas,matplotlib,seaborn,sklearn

# Project 2: Supervised Classification

This project uses python packages to perform different unsupervised learning methods on a given breast cancer and diabetes dataset.

## Part 2: Diabetes Dataset
### Importing the Libraries

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn as sk
import seaborn as sns

### Problems
1. Identify which column(s) (both train and test) has/have missing values? Identify the ‘row id’s.
“Impute” them with “Average/Most Frequent” values.

In [13]:
from sklearn.impute import KNNImputer


# importing data sets

dbt_training_data = pd.read_csv("diabetes_training.csv")
dbt_test_data = pd.read_csv("diabetes_testing.csv")

x_train = dbt_training_data.drop(columns=['id', 'class'])
y_train = dbt_training_data['class']
x_test = dbt_test_data.drop(columns=['id', 'class'])
y_test = dbt_test_data['class']

#imputer strategy to k nearest neighbors

knn_columns = ['plas','pres','skin','insu','mass','pedi','age']
imputer_knn = KNNImputer(n_neighbors=3)

x_train_knn = x_train[knn_columns].copy()
x_train_knn.replace(0, np.nan, inplace=True)
x_test_knn = x_test[knn_columns].copy()
x_test_knn.replace(0, np.nan, inplace=True)

train_imputed_knn = pd.DataFrame(imputer_knn.fit_transform(x_train_knn), columns=knn_columns)
test_imputed_knn = pd.DataFrame(imputer_knn.transform(x_test_knn), columns=knn_columns)

2. Calculate accuracy using each of these classifiers (up to 3 decimal places):

3. Now tweak the parameters of the above models, what is the best result you can get? Write the answer and upload the workbook as proof. Name this classifier widget as “<classifier>-best”. Example (if the tree widget is the best performer)

#### Logistic Regression

In [14]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

x_train = train_imputed_knn
x_test = test_imputed_knn


clf = LogisticRegression(penalty='l2', C=0.5, max_iter=1000)
clf.fit(x_train, y_train)

# Metrics
y_pred = clf.predict(x_test)
print(classification_report(y_test, y_pred, digits=3))


                 precision    recall  f1-score   support

tested_negative      0.779     0.870     0.822       146
tested_positive      0.721     0.576     0.641        85

       accuracy                          0.762       231
      macro avg      0.750     0.723     0.731       231
   weighted avg      0.758     0.762     0.755       231



In [183]:
# tweaking parameters for Logistic Regression
clf_best = LogisticRegression(penalty='l2', solver='liblinear', C=5, max_iter=1000)
clf_best.fit(x_train, y_train)
y_pred_best = clf_best.predict(x_test)
print(classification_report(y_test, y_pred_best, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.791     0.884     0.835       146
tested_positive      0.750     0.600     0.667        85

       accuracy                          0.779       231
      macro avg      0.771     0.742     0.751       231
   weighted avg      0.776     0.779     0.773       231



#### Naive Bayes

In [16]:
# Naive Bayes
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()
y_pred_gnb = gnb.fit(x_train, y_train).predict(x_test)
print(classification_report(y_test, y_pred_gnb, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.787     0.836     0.811       146
tested_positive      0.684     0.612     0.646        85

       accuracy                          0.753       231
      macro avg      0.736     0.724     0.728       231
   weighted avg      0.749     0.753     0.750       231



#### SVM

In [185]:
# SVM
from sklearn import svm

svm_clf = svm.SVC(C=1.0, kernel='rbf', gamma='auto', max_iter=100)
svm_clf.fit(x_train, y_train)
y_pred_svm = svm_clf.predict(x_test)
print(classification_report(y_test, y_pred_svm, zero_division=0, digits=3))


                 precision    recall  f1-score   support

tested_negative      0.734     0.699     0.716       146
tested_positive      0.522     0.565     0.542        85

       accuracy                          0.649       231
      macro avg      0.628     0.632     0.629       231
   weighted avg      0.656     0.649     0.652       231



/home/acenyiokopc/miniconda3/envs/sklearn/lib/python3.13/site-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [191]:

# tweaking parameters for SVM
svm_clf_best = svm.SVC(C=7, kernel='poly', gamma='scale', max_iter=-1)
svm_clf_best.fit(x_train, y_train)
y_pred_svm_best = svm_clf_best.predict(x_test)
print(classification_report(y_test, y_pred_svm_best, zero_division=0, digits=3))


                 precision    recall  f1-score   support

tested_negative      0.764     0.911     0.831       146
tested_positive      0.772     0.518     0.620        85

       accuracy                          0.766       231
      macro avg      0.768     0.714     0.725       231
   weighted avg      0.767     0.766     0.753       231



#### Random Forest

In [ ]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier
rf_clf = RandomForestClassifier(n_estimators=50, min_samples_split=10)
rf_clf.fit(x_train, y_train)
y_pred_rf = rf_clf.predict(x_test)
print(classification_report(y_test, y_pred_rf, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.778     0.890     0.831       146
tested_positive      0.750     0.565     0.644        85

       accuracy                          0.771       231
      macro avg      0.764     0.728     0.737       231
   weighted avg      0.768     0.771     0.762       231



In [127]:
# tweaking random forest parameters
rf_clf_best = RandomForestClassifier(n_estimators=5, min_samples_split=10, max_depth=3)
rf_clf_best.fit(x_train, y_train)
y_pred_rf_best = rf_clf_best.predict(x_test)
print(classification_report(y_test, y_pred_rf_best, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.784     0.822     0.803       146
tested_positive      0.667     0.612     0.638        85

       accuracy                          0.745       231
      macro avg      0.725     0.717     0.720       231
   weighted avg      0.741     0.745     0.742       231



#### k-Nearest Neighbors

In [ ]:
# k-Nearest Neighbors
from sklearn.neighbors import KNeighborsClassifier
knn_clf = KNeighborsClassifier(n_neighbors=5, metric='euclidean', weights='uniform')
knn_clf.fit(x_train, y_train)
y_pred_knn = knn_clf.predict(x_test)
print(classification_report(y_test, y_pred_knn, digits=3))


                 precision    recall  f1-score   support

tested_negative      0.755     0.801     0.777       146
tested_positive      0.618     0.553     0.584        85

       accuracy                          0.710       231
      macro avg      0.687     0.677     0.681       231
   weighted avg      0.705     0.710     0.706       231



In [163]:
# tweaking k-NN parameters
knn_clf_best = KNeighborsClassifier(n_neighbors=17, metric='euclidean', weights='distance', algorithm='auto')
knn_clf_best.fit(x_train, y_train)
y_pred_knn = knn_clf_best.predict(x_test)
print(classification_report(y_test, y_pred_knn, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.785     0.849     0.816       146
tested_positive      0.699     0.600     0.646        85

       accuracy                          0.758       231
      macro avg      0.742     0.725     0.731       231
   weighted avg      0.753     0.758     0.753       231



#### Tree

In [22]:
from sklearn.tree import DecisionTreeClassifier
tree_clf = DecisionTreeClassifier(max_depth=5, min_samples_split=5, min_samples_leaf=2)
tree_clf.fit(x_train, y_train)
y_pred_tree = tree_clf.predict(x_test)
print(classification_report(y_test, y_pred_tree, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.763     0.815     0.788       146
tested_positive      0.640     0.565     0.600        85

       accuracy                          0.723       231
      macro avg      0.701     0.690     0.694       231
   weighted avg      0.718     0.723     0.719       231



In [176]:
# tweaking tree parameters
tree_clf_best = DecisionTreeClassifier(max_depth=15, min_samples_split=5, min_samples_leaf=3)
tree_clf_best.fit(x_train, y_train)
y_pred_tree_best = tree_clf_best.predict(x_test)
print(classification_report(y_test, y_pred_tree_best, digits=3))

                 precision    recall  f1-score   support

tested_negative      0.758     0.774     0.766       146
tested_positive      0.598     0.576     0.587        85

       accuracy                          0.701       231
      macro avg      0.678     0.675     0.676       231
   weighted avg      0.699     0.701     0.700       231



### Final Result

The model with the best accuracy after parameter modification was the **Logistic Regression model** with a CA of **0.779**